exp/query/cache.go

```diff
@@ -0,0 +1,158 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package query
+
+import (
+	"container/list"
+	"errors"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+var (
+	ErrTenantNotFound       = errors.New("cache: tenant not found")
+	ErrTenantDirectoryFound = errors.New("cache: tenant directory not found")
+)
+
+const (
+	evictFraction = 0.1 // When evicting, remove 10% of max disk capacity.
+	CachePrefix   = "weaviate-cache"
+)
+
+// DiskCache is LRU disk based cache for tenants on-disk data.
+type DiskCache struct {
+	// key: <collection>/<tenant>
+	tenants   map[string]*list.Element
+	evictList *list.List
+	mu        sync.Mutex // TODO(kavi): Make it more fine-grained than global mutex
+
+	// more than this will start evicting the tenants
+	// maxCap in bytes.
+	maxCap int64
+
+	basePath string
+}
+
+func NewDiskCache(basePath string, maxCap int64) *DiskCache {
+	return &DiskCache{
+		tenants:   make(map[string]*list.Element),
+		evictList: list.New(),
+		basePath:  path.Join(basePath, CachePrefix),
+		maxCap:    maxCap,
+	}
+}
+
+// Assumes id all small-case
+func (d *DiskCache) Tenant(collection, tenantID string) (*TenantCache, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if c, exists := d.tenants[d.TenantKey(collection, tenantID)]; exists {
+		d.evictList.MoveToFront(c)
+		c.Value.(*TenantCache).LastAccessed = time.Now()
+		return c.Value.(*TenantCache), nil
+	}
+	return nil, ErrTenantNotFound
+}
+
+// AddTenant is called after having the files in right directory
+// AddTenant checks if file is present on that directory
+// deterministically generated from `collection` & `tenant` with basePath
+func (d *DiskCache) AddTenant(collection, tenantID string, version uint64) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	tc := &TenantCache{
+		TenantID:     tenantID,
+		Collection:   collection,
+		Version:      version,
+		LastAccessed: time.Now(),
+		basePath:     d.basePath,
+	}
+
+	_, err := os.Stat(tc.AbsolutePath())
+	if errors.Is(err, os.ErrNotExist) {
+		return errors.Join(ErrTenantDirectoryFound, err)
+	}
+
+	c := d.evictList.PushFront(tc)
+	d.tenants[d.TenantKey(tc.Collection, tc.TenantID)] = c
+
+	// evit if the size is getting filled up
+	// TODO(kavi): Worth doing it on different go routine? But will get complex with correctness
```

Comment:
i'm fine with leaving this usage call here for now, but as a follow up can you add some metrics for how often it's called/how long it takes?